new-frontiers-14/frontier-station-14
# Give parrots the ability to repeat learned phrases

## About the PR
This PR gives parrots (MobParrot) the ability to repeat overheard phrases or parts of phrases. It accomplishes this by doing the following:

* Adds a new component, `ParrotSpeechComponent`, which defines various parameters and keeps track of learned phrases and timings.
* Adds a new system, `ParrotSpeechSystem`, which listens for nearby utterances and makes the birds occasionally say one of their learned phrases.
* Updates `MobParrot` to include a ParrotSpeech component, as well as ActiveListener (otherwise it won't receive ListenEvents).

## Why / Balance
There is a guaranteed parrot on the Pirate's Cove, as well as one on the Crescent for whatever reason. In an attempt to inject some more life into them, why not have parrots, well, parrot the words they hear?

The various timing parameters probably have to be tweaked somewhat. I don't want parrots to talk non-stop, as it quickly gets tiring, but it also shouldn't be so rare that you never encounter a repeated phrase at all.

* With a learn chance of 20%, after hearing 10 messages, there's an ~89% chance the parrot will have learned at least one phrase, which seems acceptable.
* A parroted phrase every 2–5 minutes results in at *most* ~183 utterances in a 6-hour shift.
* With parrots being relatively rare (can't buy them at cargo), the feature is unlikely to be abused. Parrots also have a cap to the number of phrases they can learn, as well as the length (in number of words) of each phrase. Parrots do not forget old phrases over time, however.
* Phrases are randomly selected sequences of words from an overheard message. Parrots will prefer longer phrases, but can end up learning single words too. Phrases can span across sentences (e.g., "that Pioneer. Seems very") and otherwise not make sense without context (e.g., "all of the"). This is not only acceptable but desirable: the parrot has no idea what it's saying.
* The accent of the original message is retained. Parrots will faithfully recreate the hisssing ssspeech of lizardsss, the buzzz of moths, wotevah da goblins are sayin', and so on. Also intended and fun.
* Parrot messages do not show up in the chat log, to reduce spam, same as advertisements from vending machines and the like.
* Parrot messages _also_ do not show up in admin logs. If a parrot is found to be flying around Frontier saying bannably offensive things, the admin will have to find the player that taught the parrot that phrase. It is fairly trivial to add an admin log entry like "Parrot (id) learned the phrase (phrase) from (source)", if desired.
* Parrots do *not* learn from each other's utterances. If they did, they'd pick up on the various "RAWWK!" and "SQUAWK!" things the parrot accent adds, which would inevitably pollute the list of learned phrases, possibly with a fun snowball effect.
* When anything with a ParrotSpeechComponent has an active mind, the random chatter is paused.

I have no current intention to upstream this.

## How to test
1. Spawn a parrot.
2. Update its ParrotSpeechComponent to reduce the delay between messages, and set the LearnChance to 1.
3. Make sure you are a living mob near the parrot, not a ghost.
4. Speak.
5. Wait for the parrot to repeat parts of what you've said.
6. Spawn more parrots, play around. :)
7. Ensure that the parrots *don't* repeat each other's utterances.
8. Put vending machines around a parrot. Wait for them to talk. Make sure the parrot does not repeat advertisements.
9. Kill the parrots. They should cease their mindless chatter.

## Media
(TODO)


## Breaking changes
No.

**Changelog**
:cl:
- add: Parrots will now occasionally repeat some of the things you say near them. Better watch your words when you plan your next act of piracy!
